crates/ruff_python_formatter/src/expression/expr_bin_op.rs

```diff
@@ -290,3 +272,490 @@ impl NeedsParentheses for ExprBinOp {
+}
+
+impl Operator<'_> {
+    fn precedence(&self) -> OperatorPrecedence {
+        OperatorPrecedence::from(self.symbol)
+    }
+
+    fn has_trailing_comments(&self) -> bool {
+        !self.trailing_comments.is_empty()
+    }
+}
+
+impl Format<PyFormatContext<'_>> for Operator<'_> {
+    fn fmt(&self, f: &mut Formatter<PyFormatContext<'_>>) -> FormatResult<()> {
+        write!(
+            f,
+            [
+                self.symbol.format(),
+                trailing_comments(self.trailing_comments)
+            ]
+        )
+    }
+}
+
+/// Index of an Operand in the [`FlatBinaryExpressionSlice`].
+#[derive(Copy, Clone, Debug, PartialEq, Eq, Ord, PartialOrd)]
+struct OperandIndex(usize);
+
+impl OperandIndex {
+    fn new(index: usize) -> Self {
+        debug_assert_eq!(index % 2, 0, "Operand indices must be even positions");
+
+        Self(index)
+    }
+
+    /// Returns the index of the operator directly left to this operand. Returns [`None`] if
+    /// this is the first operand in the call chain.
+    fn left_operator(self) -> Option<OperatorIndex> {
+        if self.0 == 0 {
+            None
+        } else {
+            Some(OperatorIndex::new(self.0 - 1))
+        }
+    }
+
+    /// Returns the index of the operand's right operator. The method always returns an index
+    /// even if the operand has no right operator. Use [`BinaryCallChain::get_operator`] to test if
+    /// the operand has a right operator.
+    fn right_operator(self) -> OperatorIndex {
+        OperatorIndex::new(self.0 + 1)
+    }
+}
+
+/// Index of an Operator in the [`FlatBinaryExpressionSlice`].
+#[derive(Copy, Clone, Debug, PartialEq, Eq, Ord, PartialOrd)]
+struct OperatorIndex(NonZeroUsize);
+
+impl OperatorIndex {
+    fn new(index: usize) -> Self {
+        assert_eq!(index % 2, 1, "Operator indices must be odd positions");
+
+        // SAFETY A value with a module 0 is guaranteed to never equal 0
+        #[allow(unsafe_code)]
+        Self(unsafe { NonZeroUsize::new_unchecked(index) })
+    }
+
+    fn right_operand(self) -> OperandIndex {
+        OperandIndex::new(self.0.get() + 1)
+    }
+}
+
+impl<'a> Index<OperatorIndex> for FlatBinaryExpressionSlice<'a> {
+    type Output = Operator<'a>;
+
+    fn index(&self, index: OperatorIndex) -> &Self::Output {
+        self.0[index.0.get()].unwrap_operator()
+    }
+}
+
+impl<'a> Index<OperandIndex> for FlatBinaryExpressionSlice<'a> {
+    type Output = Operand<'a>;
+
+    fn index(&self, index: OperandIndex) -> &Self::Output {
+        self.0[index.0].unwrap_operand()
+    }
+}
+
+mod size_assertion {
+    use crate::expression::expr_bin_op::{
+        FlatBinaryExpressionSlice, OperandOrOperator, OperatorIndex,
+    };
+
+    static_assertions::assert_eq_size!(Option<OperatorIndex>, OperatorIndex);
+
+    static_assertions::assert_eq_size!(&FlatBinaryExpressionSlice, &[OperandOrOperator]);
+    static_assertions::assert_eq_align!(&FlatBinaryExpressionSlice, &[OperandOrOperator]);
+}
```

Comment:
Nice. This in turn ensures that the unsafe constructor is valid?

Reply:
I don't think it's a 100% guaranteed but the unsafe code is invalid for sure if these constraints are violated.